Persistence service implementation for AWS DynamoDB

ssalonen opened the pull request:
Fixes #4541 

- Tested with openhab1 and openhab2 (with compatibility layer)
- Integration tests with all item types (requires AWS credentials, see test project README)
- Documentation in Wiki: https://github.com/openhab/openhab/wiki/Amazon-DynamoDB-Persistence

Precompiled JAR: https://github.com/ssalonen/openhab/releases/tag/pooling-pr-20160517-rebase-utc-fix (matches 9e6adab)

ssalonen commented:
Hi anything needed for me to get this going forward?

Currently this is implemented for openhab1. What is recommended to make it more openhab2 friendly, or is this enough?

I guess single code base for both openhab1 and openhab2 is nice as long as both versions need to be supported.

watou commented on bundles/persistence/org.openhab.persistence.dynamodb/.classpath:
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<classpath>
+	<classpathentry exported="true" kind="lib" path="lib/aws-java-sdk-core-1.10.69.jar" sourcepath="/home/salsam/openhab2-master3/aws-java-sdk-1.10.69/lib/aws-java-sdk-1.10.69-sources.jar"/>
```
sourcepath refers to local directory.

ssalonen replied:
Fixed

ssalonen commented:
Btw, current implementation does not persist Undefined values. I took this approach from InfluxDB but just want to know what is expected from a persistence binding?

It seems db4o persistence actually stores undefined values as well -- at least as far I can analyze the code

watou commented:
I believe it is correct to not persist Undefined states.  My interpretation is that Undefined means the absence of information, and persisting an item state should only happen when the state carries information.  Why db4o would persist Undefined states seems like a bug.

ssalonen commented:
Two months passed now since last comment. Anything I can do to progress this PR?

watou commented:
From my side, I only lack sufficient time to dig into a proper review.  But other maintainers are more than welcome to contribute their own.  If @9037568, @lewie or any other experienced person would like to contribute a review, that would be very helpful.

Reading some of the old comments regarding persisting Undefined states, I've read that there is a proposed reference implementation of a persistence service in ESH that would persist all state changes, regardless of state type (including UNDEF, NULL, etc.).  It makes sense to persist all state changes as long as clients can query what they want.  Perhaps in the past it has been up to an openHAB persistence service author's discretion which data types to persist from the item type's accepted data types.  

But all that is in place now is the OH1-style persistence service, so I think your contribution is very worthwhile, regardless of the above.

9037568 commented:
I've begun a review. Is there a document that describes how this component is meant to be configured/used?

ssalonen commented:
@9037568 there is link to documentation wiki in the first message

9037568 commented on bundles/persistence/org.openhab.persistence.dynamodb.test/src/test/java/org/openhab/persistence/dynamodb/internal/DynamoDBConfigTest.java:
```diff
@@ -0,0 +1,147 @@
+
+    @Test
+    public void testRegionWithAccessKeys() throws Exception {
+        DynamoDBConfig fromConfig = DynamoDBConfig.fromConfig(ImmutableMap.<String, Object> of("region", "eu-west-1",
```
This file needs to be run through the eclipse formatter.

ssalonen commented on bundles/persistence/org.openhab.persistence.dynamodb.test/src/test/java/org/openhab/persistence/dynamodb/internal/AbstractTwoItemIntegrationTest.java:
```diff
@@ -0,0 +1,334 @@
+/**
+ * See DimmerItemIntegrationTest for example how to use this base class.
+ *
+ * Implementor need to store two items. The first one needs to be smaller than the second. Protected static timestamps
```
Improved javadoc comment in a26baff55bc309b691833e9572c06e8099df064b

ssalonen commented on bundles/persistence/org.openhab.persistence.dynamodb.test/src/test/java/org/openhab/persistence/dynamodb/internal/AbstractTwoItemIntegrationTest.java:
```diff
@@ -0,0 +1,334 @@
+    /**
+     * State of the time item stored first, should be smaller than the second value
+     *
+     * @return
```
Fixed in b355205f23e3a588bedd6e3c46fb5128ebfe1ddc

ssalonen commented on bundles/persistence/org.openhab.persistence.dynamodb.test/src/test/java/org/openhab/persistence/dynamodb/internal/AbstractTwoItemIntegrationTest.java:
```diff
@@ -0,0 +1,334 @@
+    /**
+     * State of the time item stored second, should be larger than the first value
+     *
+     * @return
```
Fixed in b355205f23e3a588bedd6e3c46fb5128ebfe1ddc

ssalonen commented on bundles/persistence/org.openhab.persistence.dynamodb.test/src/test/java/org/openhab/persistence/dynamodb/internal/ContactItemIntegrationTest.java:
```diff
@@ -0,0 +1,64 @@
+    private static final String name = "contact";
+    private static final OpenClosedType state1 = OpenClosedType.CLOSED;
+    private static final OpenClosedType state2 = OpenClosedType.OPEN;
+    private static final OnOffType stateBetween = null; // no such that exists
```
Improved the comment in a3b6d523381acd1a93e5ecf11d483b5ff78cdb00

ssalonen commented on bundles/persistence/org.openhab.persistence.dynamodb/java/org/openhab/persistence/dynamodb/internal/DynamoDBClient.java:
```diff
@@ -0,0 +1,60 @@
+import com.amazonaws.services.dynamodbv2.AmazonDynamoDBClient;
+import com.amazonaws.services.dynamodbv2.document.DynamoDB;
+
+/***
```
Fixed in 2f786e314adc9450a30d2a98f51ebb5410d9c692

ssalonen commented on bundles/persistence/org.openhab.persistence.dynamodb/java/org/openhab/persistence/dynamodb/internal/DynamoDBPersistenceService.java:
```diff
@@ -0,0 +1,444 @@
+    }
+
+    public void activate(final BundleContext bundleContext, final Map<String, Object> config) {
+        logger.debug("dynamodb persistence service activated");
```
Fixed in 9739bcb2a68b881826a78e75c8f113fe5fca4b5a

ssalonen commented on bundles/persistence/org.openhab.persistence.dynamodb/java/org/openhab/persistence/dynamodb/internal/DynamoDBPersistenceService.java:
```diff
@@ -0,0 +1,444 @@
+        try {
+            dbConfig = DynamoDBConfig.fromConfig(config);
+            if (dbConfig == null) {
+                throw new IllegalArgumentException("Something was wrong with configuration");
```
If you look the implementation of the static method fromConfig, you can find out the cases when it returns null. These include missing AWS region (this is error), and missing configuration parameters. 

So I do think this is an error with configuration.

But what does it mean if `config`  is `null`? This binding has mandatory parameters and cannot work without configuration. How should I handle that?

9037568 replied:
If the ```config``` object that the ```fromConfig``` method receives is null, that should mean that there were no dynamodb settings in the openhab configuration file. I would think the ```fromConfig``` method should check for this, and if true, log an error and return null to the caller.

If ```fromConfig``` logs appropriately, then the caller doesn't need to log anything.  If it receives a null response, it simply returns.

ssalonen replied:
Makes sense. Fixed in fe6b1de. 

Added tests for null configuration and empty configuration. Also added test for invalid profile file.

ssalonen commented on bundles/persistence/org.openhab.persistence.dynamodb/java/org/openhab/persistence/dynamodb/internal/DynamoDBPersistenceService.java:
```diff
@@ -0,0 +1,444 @@
+                // No table present, continue with creation
+                db.getDynamoClient().createTable(request);
+            } catch (AmazonClientException e) {
+                logger.error("Exception when creation table (descibe): {}", e.getMessage());
```
Fixed in 9f5dbae2106407e35f11f0908f892a6d2b6b3667

ssalonen commented on bundles/persistence/org.openhab.persistence.dynamodb/java/org/openhab/persistence/dynamodb/internal/DynamoDBPersistenceService.java:
```diff
@@ -0,0 +1,444 @@
+                item = itemRegistry.getItem(itemName);
+            }
+        } catch (ItemNotFoundException e1) {
+            logger.error("Unable to get item type for {}", itemName);
```
You are right. Fixed in 9f5dbae2106407e35f11f0908f892a6d2b6b3667

ssalonen commented on bundles/persistence/org.openhab.persistence.dynamodb/java/org/openhab/persistence/dynamodb/internal/DynamoDBPersistenceService.java:
```diff
@@ -0,0 +1,444 @@
+        } catch (ItemNotFoundException e1) {
+            logger.error("Unable to get item type for {}", itemName);
+            // Set type to null - data will be returned as StringType
+            item = null;
```
Fixed in 9f5dbae2106407e35f11f0908f892a6d2b6b3667

ssalonen commented:
Added try-catch for queries in commit 51987b659b006396790431dbb7bfe055544999f5, noticed that despite all the check made in the persistence service, AWS can return 400 (resource/table does not exist) on the very first queries.

ssalonen commented:
Tried to connection retry (with commit b7a7e06) using the following procedure
1. have openhab configuration writing / querying regularly to/from the persistence
2. disable internet connection
3. start openhab. You will get errors since connection to AWS is down. The persistence service logs `Failed to establish the dynamodb database connection. Connection will be retried later on persistence service query/store.`
4. enable internet connection
5. writes / queries go through -- connection is recovered.

So should be OK from that part.

ssalonen commented:
@author  added to java docs. 

Anything else we need to fix?

ssalonen commented:
Could you please help me how to diagnose the Jenkins error? It gives me now 404, can I retrigger build?

9037568 commented:
Not sure what that means. I'm getting a similar 404 for the [last successful build too.](https://github.com/openhab/openhab/pull/4693)

I don't think we can trigger a build manually.

ssalonen commented on bundles/persistence/org.openhab.persistence.dynamodb/java/org/openhab/persistence/dynamodb/internal/DynamoDBPersistenceService.java:
```diff
@@ -0,0 +1,444 @@
+
+/**
+ * This is the implementation of the DynamoDB {@link PersistenceService}. It persists item values
+ * using the <a href="https://aws.amazon.com/dynamodb/">Amazon DynamoDB</a> time series database. The states (
```
Fixed in a85e536

9037568 commented:
Thanks for this contribution, @ssalonen. Very much appreciate your work on the new plugin and associated changes.

9037568 commented:
Oh, and please update the wiki page to make your docs visible!

kaikreuzer commented:
This PR causes a problem on the build, see https://openhab.ci.cloudbees.com/job/openHAB1-Addons/org.openhab.persistence$org.openhab.persistence.dynamodb.test/1314/changes.

Could you please have a look?

watou commented:
@ssalonen, here is the part of the console output:

```
09:22:32.681  |  [ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:0.24.0:compile (default-compile) on project org.openhab.persistence.dynamodb.test: Compilation failure: Compilation failure:
09:22:32.682  |  [ERROR] /scratch/jenkins/workspace/openHAB1-Addons/bundles/persistence/org.openhab.persistence.dynamodb.test/src/test/java/org/openhab/persistence/dynamodb/internal/AbstractDynamoDBItemSerializationTest.java:[15]
09:22:32.682  |  [ERROR] import java.time.ZoneId;
09:22:32.682  |  [ERROR] ^^^^^^^^^
09:22:32.682  |  [ERROR] The import java.time cannot be resolved
09:22:32.682  |  [ERROR] /scratch/jenkins/workspace/openHAB1-Addons/bundles/persistence/org.openhab.persistence.dynamodb.test/src/test/java/org/openhab/persistence/dynamodb/internal/AbstractDynamoDBItemSerializationTest.java:[16]
09:22:32.683  |  [ERROR] import java.time.ZoneOffset;
09:22:32.683  |  [ERROR] ^^^^^^^^^
09:22:32.683  |  [ERROR] The import java.time cannot be resolved
09:22:32.683  |  [ERROR] /scratch/jenkins/workspace/openHAB1-Addons/bundles/persistence/org.openhab.persistence.dynamodb.test/src/test/java/org/openhab/persistence/dynamodb/internal/AbstractDynamoDBItemSerializationTest.java:[153]
09:22:32.683  |  [ERROR] calendar.setTimeZone(TimeZone.getTimeZone(ZoneId.ofOffset("UTC", ZoneOffset.ofHours(3))));
09:22:32.683  |  [ERROR] ^^^^^^
09:22:32.684  |  [ERROR] ZoneId cannot be resolved
09:22:32.684  |  [ERROR] /scratch/jenkins/workspace/openHAB1-Addons/bundles/persistence/org.openhab.persistence.dynamodb.test/src/test/java/org/openhab/persistence/dynamodb/internal/AbstractDynamoDBItemSerializationTest.java:[153]
09:22:32.684  |  [ERROR] calendar.setTimeZone(TimeZone.getTimeZone(ZoneId.ofOffset("UTC", ZoneOffset.ofHours(3))));
09:22:32.684  |  [ERROR] ^^^^^^^^^^
09:22:32.684  |  [ERROR] ZoneOffset cannot be resolved
09:22:32.685  |  [ERROR] /scratch/jenkins/workspace/openHAB1-Addons/bundles/persistence/org.openhab.persistence.dynamodb.test/src/test/java/org/openhab/persistence/dynamodb/internal/AbstractDynamoDBItemSerializationTest.java:[166]
09:22:32.685  |  [ERROR] calendar.setTimeZone(TimeZone.getTimeZone(ZoneId.ofOffset("UTC", ZoneOffset.ofHours(3))));
09:22:32.685  |  [ERROR] ^^^^^^
09:22:32.685  |  [ERROR] ZoneId cannot be resolved
09:22:32.685  |  [ERROR] /scratch/jenkins/workspace/openHAB1-Addons/bundles/persistence/org.openhab.persistence.dynamodb.test/src/test/java/org/openhab/persistence/dynamodb/internal/AbstractDynamoDBItemSerializationTest.java:[166]
09:22:32.686  |  [ERROR] calendar.setTimeZone(TimeZone.getTimeZone(ZoneId.ofOffset("UTC", ZoneOffset.ofHours(3))));
09:22:32.686  |  [ERROR] ^^^^^^^^^^
09:22:32.686  |  [ERROR] ZoneOffset cannot be resolved
09:22:32.686  |  [ERROR] /scratch/jenkins/workspace/openHAB1-Addons/bundles/persistence/org.openhab.persistence.dynamodb.test/src/test/java/org/openhab/persistence/dynamodb/internal/AbstractDynamoDBItemSerializationTest.java:[175]
09:22:32.686  |  [ERROR] String expected = String.join(",", point.getLatitude().toBigDecimal().toString(),
09:22:32.686  |  [ERROR] ^^^^
09:22:32.687  |  [ERROR] The method join(String, String, String, String) is undefined for the type String
09:22:32.687  |  [ERROR] 7 problems (7 errors)
09:22:32.687  |  [ERROR] -> [Help 1]
09:22:32.687  |  [ERROR] 
09:22:32.687  |  [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
09:22:32.687  |  [ERROR] Re-run Maven using the -X switch to enable full debug logging.
09:22:32.688  |  [ERROR] 
09:22:32.688  |  [ERROR] For more information about the errors and possible solutions, please read the following articles:
09:22:32.688  |  [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
09:22:32.688  |  [ERROR] 
09:22:32.688  |  [ERROR] After correcting the problems, you can resume the build with the command
09:22:32.689  |  [ERROR]   mvn <goals> -rf :org.openhab.persistence.dynamodb.test
```

kaikreuzer commented:
Looks like you are using Java8 (java.time), but the code has to be Java7 compatible.

9037568 commented:
Since the compiler target is set to 1.7 in the top-level pom, shouldn't
maven throw an error during compile when it encounters 1.8-only features?

kaikreuzer commented:
Well, it throws a compilation error, which is fair because it sets the compiler to 1.7. It does not "know" that java.time is Java8, so "cannot be resolved" is a decent way to complain, I would think :-)

ssalonen commented:
Thank you for accepting this PR.

- new PR introduced to fix the Java7 fixes. I think really need to remember to test with java 7 as well next time.
- Updated [Persistence](https://github.com/openhab/openhab/wiki/Persistence) page to have the DynamoDB in the table. 
- Removed the "not available yet" disclaimer from [DynamoDB page](https://github.com/openhab/openhab/wiki/Amazon-DynamoDB-Persistence)
